StackOverflow caused by recursive uses declaration

Today I did battle with mapstruct because I accidentally implemented a class with a self referencing "uses" declarations.

```
...
@Mapper(config = MapStructConfig.class //
		, uses = TodoMapper.class //
)
public interface TodoMapper {
    public TodoResponse toResponse(TodoJpa jpa);
...

```

I do realise this is my mistake and not a functional issue with mapstruct.
However if possible it would have been nice to have an error explaining my fault, rather than a stacktrace like the following:

`[ERROR] .../TodoMapper.java:[12,8] Internal error in the mapping processor: java.lang.StackOverflowError   at com.sun.tools.javac.code.Type.hasTag(Type.java:112)          at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1967)    at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1955)    at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.asSuper(Types.java:1952)      at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1968)    at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1955)    at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.asSuper(Types.java:1952)      at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1975)    at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1955)    at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.asSuper(Types.java:1952)      at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1975)    at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1955)    at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.asSuper(Types.java:1952)      at com.sun.tools.javac.code.Types$4.visitClassType(Types.java:921)      at com.sun.tools.javac.code.Types$4.visitClassType(Types.java:844)      at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.isSubtype(Types.java:840)     at com.sun.tools.javac.code.Types.isSubtype(Types.java:804)     at com.sun.tools.javac.model.JavacTypes.isSubtype(JavacTypes.java:98)   at org.mapstruct.ap.internal.util.workarounds.SpecificCompilerWorkarounds.isSubtype(SpecificCompilerWorkarounds.java:63)        at org.mapstruct.ap.internal.util.workarounds.TypesDecorator.isSubtype(TypesDecorator.java:55)          at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:192)     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:182)     at org.mapstruct.ap.internal.model.common.TypeFactory.getParameters(TypeFactory.java:392)       at org.mapstruct.ap.internal.model.common.TypeFactory.getParameters(TypeFactory.java:377)       at org.mapstruct.ap.internal.model.common.TypeFactory.getSingleParameter(TypeFactory.java:368)          at org.mapstruct.ap.internal.model.common.Type.determineTargetType(Type.java:620)       at org.mapstruct.ap.internal.model.common.Type.getPropertyWriteAccessors(Type.java:570)         at org.mapstruct.ap.internal.model.source.TargetReference$BuilderFromTargetMapping.getTargetEntries(TargetReference.java:190)   at org.mapstruct.ap.internal.model.source.TargetReference$BuilderFromTargetMapping.build(TargetReference.java:158)      at org.mapstruct.ap.internal.model.source.Mapping.init(Mapping.java:446)        at org.mapstruct.ap.internal.model.source.Mapping.init(Mapping.java:413)        at org.mapstruct.ap.internal.model.source.SourceMethod$Builder.build(SourceMethod.java:200)     at org.mapstruct.ap.internal.processor.MM
ethodRetrievalProcessor.getMethodRequiringImplementation(MethodRetrievalProcessor.java:261)     at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.getMethod(MethodRetrievalProcessor.java:188)    at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.retrieveMethods(MethodRetrievalProcessor.java:143)      at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.retrieveMethods(MethodRetrievalProcessor.java:158)      at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.retrieveMethods(MethodRetrievalProcessor.java:158)      at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.retrieveMethods(MethodRetrievalProcessor.java:158)      at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.retrieveMethods(MethodRetrievalProcessor.java:158)      at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.retrieveMethods(MethodRetrievalProcessor.java:158)      at 
...`
Thanks, Andrew

--- a/processor/src/main/java/org/mapstruct/ap/internal/processor/MethodRetrievalProcessor.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/processor/MethodRetrievalProcessor.java
@@ -176,11 +176,22 @@ public class MethodRetrievalProcessor implements ModelElementProcessor<Void, Lis
         //Add all methods of used mappers in order to reference them in the aggregated model
         if ( usedMapper.equals( mapperToImplement ) ) {
             for ( DeclaredType mapper : mapperOptions.uses() ) {
-                methods.addAll( retrieveMethods(
-                    asTypeElement( mapper ),
-                    mapperToImplement,
-                    mapperOptions,
-                    prototypeMethods ) );
+                TypeElement usesMapperElement = asTypeElement( mapper );
+                if ( !mapperToImplement.equals( usesMapperElement ) ) {
+                    methods.addAll( retrieveMethods(
+                        usesMapperElement,
+                        mapperToImplement,
+                        mapperOptions,
+                        prototypeMethods ) );
+                }
+                else {
+                    messager.printMessage(
+                        mapperToImplement,
+                        mapperOptions.getAnnotationMirror(),
+                        Message.RETRIEVAL_MAPPER_USES_CYCLE,
+                        mapperToImplement
+                    );
+                }
             }
         }
 

--- a/processor/src/main/java/org/mapstruct/ap/internal/util/Message.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/util/Message.java
@@ -149,6 +149,7 @@ public enum Message {
     RETRIEVAL_WILDCARD_SUPER_BOUND_SOURCE( "Can't generate mapping method for a wildcard super bound source." ),
     RETRIEVAL_WILDCARD_EXTENDS_BOUND_RESULT( "Can't generate mapping method for a wildcard extends bound result." ),
     RETRIEVAL_CONTEXT_PARAMS_WITH_SAME_TYPE( "The types of @Context parameters must be unique." ),
+    RETRIEVAL_MAPPER_USES_CYCLE( "The mapper %s is referenced itself in Mapper#uses.", Diagnostic.Kind.WARNING ),
 
     INHERITINVERSECONFIGURATION_DUPLICATES( "Several matching inverse methods exist: %s(). Specify a name explicitly." ),
     INHERITINVERSECONFIGURATION_INVALID_NAME( "None of the candidates %s() matches given name: \"%s\"." ),

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2185/Issue2185Test.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2185/Issue2185Test.java
@@ -1,0 +1,45 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2185;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.compilation.annotation.CompilationResult;
+import org.mapstruct.ap.testutil.compilation.annotation.Diagnostic;
+import org.mapstruct.ap.testutil.compilation.annotation.ExpectedCompilationOutcome;
+import org.mapstruct.ap.testutil.runner.AnnotationProcessorTestRunner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Filip Hrisafov
+ */
+@IssueKey("2185")
+@RunWith(AnnotationProcessorTestRunner.class)
+@WithClasses({
+    TodoMapper.class
+})
+public class Issue2185Test {
+
+    @Test
+    @ExpectedCompilationOutcome(value = CompilationResult.SUCCEEDED,
+        diagnostics = {
+            @Diagnostic(type = TodoMapper.class,
+                kind = javax.tools.Diagnostic.Kind.WARNING,
+                line = 14,
+                message = "The mapper org.mapstruct.ap.test.bugs._2185.TodoMapper is referenced itself in Mapper#uses.")
+        }
+    )
+    public void shouldCompile() {
+
+        TodoMapper.TodoResponse response = TodoMapper.INSTANCE.toResponse( new TodoMapper.TodoEntity( "test" ) );
+
+        assertThat( response ).isNotNull();
+        assertThat( response.getNote() ).isEqualTo( "test" );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_2185/TodoMapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_2185/TodoMapper.java
@@ -1,0 +1,46 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2185;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+/**
+ * @author Filip Hrisafov
+ */
+@Mapper(uses = TodoMapper.class)
+public interface TodoMapper {
+
+    TodoMapper INSTANCE = Mappers.getMapper( TodoMapper.class );
+
+    TodoResponse toResponse(TodoEntity entity);
+
+    class TodoResponse {
+
+        private final String note;
+
+        public TodoResponse(String note) {
+            this.note = note;
+        }
+
+        public String getNote() {
+            return note;
+        }
+    }
+
+    class TodoEntity {
+
+        private final String note;
+
+        public TodoEntity(String note) {
+            this.note = note;
+        }
+
+        public String getNote() {
+            return note;
+        }
+    }
+}

